allennlp/common/file_utils.py

```diff
@@ -85,34 +88,96 @@ def filename_to_url(filename: str, cache_dir: str = None) -> Tuple[str, str]:
     return url, etag
 
 
-def cached_path(url_or_filename: Union[str, Path], cache_dir: str = None) -> str:
+def cached_path(
+    url_or_filename: Union[str, Path],
+    cache_dir: Union[str, Path] = None,
+    extract_archive: bool = False,
+    force_extract: bool = False,
+) -> str:
     """
     Given something that might be a URL (or might be a local path),
     determine which. If it's a URL, download the file and cache it, and
     return the path to the cached file. If it's already a local path,
     make sure the file exists and then return the path.
+
+    # Parameters
+
+    url_or_filename : `Union[str, Path]`
+        A URL or local file to parse and possibly download.
+
+    cache_dir : `Union[str, Path]`, optional (default = `None`)
+        The directory to cache downloads.
+
+    extract_archive : `bool`, optional (default = `False`)
+        If `True`, then zip or tar.gz archives will be automatically extracted.
+        In which case the directory is returned.
+
+    force_extract : `bool`, optional (default = `False`)
+        If `True` and the file is an archive file, it will be extracted regardless
+        of whether or not the extracted directory already exists.
     """
     if cache_dir is None:
         cache_dir = CACHE_DIRECTORY
+
     if isinstance(url_or_filename, Path):
         url_or_filename = str(url_or_filename)
 
     url_or_filename = os.path.expanduser(url_or_filename)
     parsed = urlparse(url_or_filename)
 
+    file_path: str
+    extraction_path: Optional[str] = None
+
     if parsed.scheme in ("http", "https", "s3"):
         # URL, so get it from the cache (downloading if necessary)
-        return get_from_cache(url_or_filename, cache_dir)
+        file_path = get_from_cache(url_or_filename, cache_dir)
+
+        if extract_archive and (is_zipfile(file_path) or tarfile.is_tarfile(file_path)):
```

Comment:
Seems like we should pull this out into a separate, small pip utility, that can be shared by both us and huggingface.